azkaban-common/src/main/java/azkaban/executor/ExecutionController.java

```diff
@@ -489,8 +505,71 @@ public void retryFailures(final ExecutableFlow exFlow, final String userId)
   @Override
   public String submitExecutableFlow(final ExecutableFlow exflow, final String userId)
       throws ExecutorManagerException {
-    // Todo: insert the execution to DB queue
-    return null;
+    final String exFlowKey = exflow.getProjectName() + "." + exflow.getId() + ".submitFlow";
+    // Use project and flow name to prevent race condition when same flow is submitted by API and
+    // schedule at the same time
+    // causing two same flow submission entering this piece.
+    synchronized (exFlowKey.intern()) {
+      final String flowId = exflow.getFlowId();
+      logger.info("Submitting execution flow " + flowId + " by " + userId);
+
+      String message = "";
+
+      final int projectId = exflow.getProjectId();
+      exflow.setSubmitUser(userId);
+      exflow.setSubmitTime(System.currentTimeMillis());
+
+      final List<Integer> running = getRunningFlows(projectId, flowId);
+
+      ExecutionOptions options = exflow.getExecutionOptions();
+      if (options == null) {
+        options = new ExecutionOptions();
+      }
+
+      if (options.getDisabledJobs() != null) {
+        FlowUtils.applyDisabledJobs(options.getDisabledJobs(), exflow);
+      }
+
+      if (!running.isEmpty()) {
+        if (running.size() > this.maxConcurrentRunsOneFlow) {
+          throw new ExecutorManagerException("Flow " + flowId
+              + " has more than " + this.maxConcurrentRunsOneFlow + " concurrent runs. Skipping",
+              ExecutorManagerException.Reason.SkippedExecution);
+        } else if (options.getConcurrentOption().equals(
+            ExecutionOptions.CONCURRENT_OPTION_PIPELINE)) {
+          Collections.sort(running);
+          final Integer runningExecId = running.get(running.size() - 1);
+
+          options.setPipelineExecutionId(runningExecId);
+          message =
+              "Flow " + flowId + " is already running with exec id "
+                  + runningExecId + ". Pipelining level "
+                  + options.getPipelineLevel() + ". \n";
+        } else if (options.getConcurrentOption().equals(
+            ExecutionOptions.CONCURRENT_OPTION_SKIP)) {
+          throw new ExecutorManagerException("Flow " + flowId
+              + " is already running. Skipping execution.",
+              ExecutorManagerException.Reason.SkippedExecution);
+        } else {
+          message =
+              "Flow " + flowId + " is already running with exec id "
+                  + StringUtils.join(running, ",")
+                  + ". Will execute concurrently. \n";
+        }
+      }
+
+      final boolean memoryCheck =
+          !ProjectWhitelist.isProjectWhitelisted(exflow.getProjectId(),
+              ProjectWhitelist.WhitelistType.MemoryCheck);
+      options.setMemoryCheck(memoryCheck);
+
+      // The exflow id is set by the loader. So it's unavailable until after
+      // this call.
+      this.executorLoader.uploadExecutableFlow(exflow);
```

Comment:
Most of the implementations remain the same as those in `ExecutorManager`. The only difference is that we are not enqueueing to `queueFlows` cache now and not adding new executions to `active_executing_flows` DB table which has been deprecated already.